### PR TITLE
Add support for ON CONFLICT DO UPDATE for compressed hypertables.

### PR DESCRIPTION
**Pull request by sb230132**
This patch fixes execution of INSERT with ON CONFLICT DO UPDATE by removing error and allowing UPDATE do happen on the given compressed hypertable.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ accidentally triggering the load of a previous DB version.**
 * #5361 Add parallel support for partialize_agg()
 * #5252 Improve unique constraint support on compressed hypertables
 * #5312 Add timeout support to ping_data_node()
+* #5454 Fix ON CONFLICT DO UPDATE on compressed hypertables
```
Comment (svenklemm) [inline CHANGELOG.md:15]:
```suggestion
* #5454 Add support for ON CONFLICT DO UPDATE for compressed hypertables
```

**Review by svenklemm — COMMENTED** [review]
The PR looks good but I think we want some more tests here.

File: CHANGELOG.md (hunk 1 above)
Comment (sb230132) [reply CHANGELOG.md]: Fixed.

**Comment by sb230132** [comment]
> The PR looks good but I think we want some more tests here.

Added few more tests.

File: src/nodes/chunk_dispatch/chunk_dispatch.c [hunk 2]
```diff
@@ -145,7 +146,14 @@ ts_chunk_dispatch_get_chunk_insert_state(ChunkDispatch *dispatch, Point *point,
 			 * postgres can do proper constraint checking.
 			 */
 			if (ts_cm_functions->decompress_batches_for_insert)
+			{
 				ts_cm_functions->decompress_batches_for_insert(cis, chunk, slot);
+				OnConflictAction onconflict_action =
+					chunk_dispatch_get_on_conflict_action(dispatch);
+				/* mark rows visible */
+				if (onconflict_action == ONCONFLICT_UPDATE)
+					dispatch->estate->es_output_cid = GetCurrentCommandId(false);
```
Comment (kgyrtkirk) [inline src/nodes/chunk_dispatch/chunk_dispatch.c:155]: I wonder - how did this worked before for the normal `insert` path? why didn't it needed `es_output_id`  before? - or for normal inserts this is not "needed" ?

I have feel that its a bit too late here to set an estate field - I think this field should be set way-way-back..but I'm not sure where to place it...

[the remark this answers is not in the document]
Comment (sb230132) [reply src/nodes/chunk_dispatch/chunk_dispatch.c:155]: Ok, fixed.

Comment (sb230132) [reply src/nodes/chunk_dispatch/chunk_dispatch.c:155]: Frankly i don't know either, i face an error as "ERROR:  attempted to update invisible tuple"
To overcome this error i had to set es_output_cid to latest current command ID.

Comment (kgyrtkirk) [reply src/nodes/chunk_dispatch/chunk_dispatch.c]: it should fix that for sure - and with `true` passed I think its okayish

I'll try to keep this in mind - and let you know if I've found a better place to set this...

**Review by kgyrtkirk — APPROVED** [review]

File: src/nodes/chunk_dispatch/chunk_dispatch.c (hunk 2 above)
Comment (konskov) [reply src/nodes/chunk_dispatch/chunk_dispatch.c]: Looking at `ExecOnConflictUpdate()` , for the case TM_Invisible, where the error comes from, I see that 
```
test = table_tuple_lock(relation, conflictTid,
							estate->es_snapshot,
							existing, estate->es_output_cid,
							lockmode, LockWaitBlock, 0,
							&tmfd);
```
uses `estate->es_snapshot` and `estate->es_output_cid`, so I also think it makes sense to do this here.

**Review by konskov — APPROVED** [review]

File: src/nodes/chunk_dispatch/chunk_dispatch.c (hunk 2 above)
Comment (sb230132) [reply src/nodes/chunk_dispatch/chunk_dispatch.c]: Thanks.

**Pull request merged**